tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,711 @@
+		if int64(maxt) < mint {
+			removedFiles = append(removedFiles, seq)
+		}
+	}
+	w.bsMtx.RUnlock()
+
+	return w.deleteFiles(removedFiles)
+}
+func (w *HeadReadWriter) deleteFiles(removedFiles []int) error {
+	closers := make([]io.Closer, 0, len(removedFiles))
+	w.bsMtx.Lock()
+	for _, seq := range removedFiles {
+		w.size -= int64(w.bs[seq].Len())
+		closers = append(closers, w.cs[seq])
+		delete(w.bs, seq)
+		delete(w.cs, seq)
+		if err := os.Remove(segmentFile(w.dirFile.Name(), seq)); err != nil {
+			w.bsMtx.Unlock()
+			var merr tsdb_errors.MultiError
+			merr.Add(err)
+			merr.Add(closeAll(closers))
+			return merr.Err()
+		}
+	}
+	w.bsMtx.Unlock()
+
+	return closeAll(closers)
+}
+
+func (w *HeadReadWriter) Close() error {
+	// 'WriteChunk' locks wbufLock first and then bsMtx for cutting segment.
+	// The lock order should not be reversed here else it can cause deadlocks.
+	w.wbufLock.Lock()
+	defer w.wbufLock.Unlock()
+	w.bsMtx.Lock()
+	defer w.bsMtx.Unlock()
+
+	select {
+	case <-w.quit:
+		return nil
+	default:
+	}
+
+	close(w.quit)
+
+	if err := w.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	if err := w.dirFile.Close(); err != nil {
+		return err
+	}
+
+	return closeAllFromMap(w.cs)
+}
+
+func (w *HeadReadWriter) flushBuffer() error {
+	// Assumes that wbufLock is locked before calling this method.
+	if err := w.wbuf.Flush(); err != nil {
+		return nil
+	}
+	for i := 0; i < inBufferShards; i++ {
+		w.inBufferChunksLocks[i].Lock()
+	}
+	for i := 0; i < inBufferShards; i++ {
+		w.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
+	}
+	for i := inBufferShards - 1; i >= 0; i-- {
+		w.inBufferChunksLocks[i].Unlock()
+	}
+	return nil
+}
+
+func (w *HeadReadWriter) getChunkFromBuffer(chunkRef uint64) chunkenc.Chunk {
+	shardIdx := chunkRef % inBufferShards
+
+	w.inBufferChunksLocks[shardIdx].Lock()
+	defer w.inBufferChunksLocks[shardIdx].Unlock()
+
+	return w.inBufferChunks[shardIdx][chunkRef]
+}
+
+func (w *HeadReadWriter) chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+func (w *HeadReadWriter) shouldCutSegment(chunkLength int, maxt int64) bool {
+	return w.n == 0 || // First segment
+		(maxt-w.curFileMint > w.segmentTime && w.n > HeadSegmentHeaderSize) || // Time duration reached for the existing file.
+		w.n+int64(chunkLength+27) >= MaxSegmentSize
+}
+
+func (w *HeadReadWriter) seq() int {
+	return w.curFileSequence
+}
+
+func (w *HeadReadWriter) write(b []byte) error {
+	n, err := w.wbuf.Write(b)
+	atomic.AddInt64(&w.n, int64(n))
+	return err
+}
+
+func (w *HeadReadWriter) cut(mint int64) (returnErr error) {
+	// Sync current tail to disk and close.
+	if err := w.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(w.dirFile, MagicHeadChunks, headChunksFormatV1, 0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// The file should not be closed if there is no error,
+		// its kept open in the HeadReadWriter.
+		if returnErr != nil {
+			f.Close()
```

Comment:
We need the new file to be open as it will be the one being appended to. So we don't close inside `cutSegmentFile`. (closing and reopening is additional latency in the hot path)

Reply:
yep, my bad